packages/vite-plugin-cloudflare/src/node-js-compat.ts

```diff
@@ -1,18 +1,124 @@
-import { createRequire } from "node:module";
+import assert from "node:assert";
+import { cloudflare } from "@cloudflare/unenv-preset";
 import MagicString from "magic-string";
 import { getNodeCompat } from "miniflare";
-import * as unenv from "unenv";
-import type { WorkerConfig } from "./plugin-config";
-import type { Environment } from "vite";
+import { defineEnv } from "unenv";
+import { resolvePluginConfig } from "./plugin-config";
+import type {
+	PluginConfig,
+	ResolvedPluginConfig,
+	WorkerConfig,
+} from "./plugin-config";
+import type { Plugin } from "vite";
+
+/**
+ * A Vite plugin that can provide Node.js compatibility support for Vite Environments that are hosted in Cloudflare Workers.
+ */
+export function nodejsCompatPlugin(pluginConfig: PluginConfig): Plugin {
+	let resolvedPluginConfig: ResolvedPluginConfig;
+	return {
+		name: "vite-plugin-nodejs-compat",
+		config(userConfig, env) {
+			// Capture the configuration of the Cloudflare Workers for use in other hooks, below.
+			resolvedPluginConfig = resolvePluginConfig(pluginConfig, userConfig, env);
+
+			// Configure Vite with the Node.js polyfill aliases
+			// We have to do this across the whole Vite config because it is not possible to do it per Environment.
+			return {
+				resolve: {
+					alias: getNodeCompatAliases(),
+				},
+			};
+		},
+		configEnvironment(environmentName) {
+			if (resolvedPluginConfig.type === "assets-only") {
+				return;
+			}
+
+			const workerConfig = resolvedPluginConfig.workers[environmentName];
+			if (!workerConfig) {
+				return;
+			}
+
+			if (!isNodeCompat(workerConfig)) {
+				return;
+			}
+
+			// Ignore the Node.js external modules when building.
+			return {
+				build: {
+					rollupOptions: {
+						external: getNodeCompatExternals(),
+					},
+				},
+			};
+		},
+		async resolveId(source) {
+			if (resolvedPluginConfig.type === "assets-only") {
+				return;
+			}
+
+			const workerConfig = resolvedPluginConfig.workers[this.environment.name];
+			if (!workerConfig) {
+				return;
+			}
+
+			const unresolvedAlias = dealiasVirtualNodeJSImports(source, workerConfig);
+			if (!unresolvedAlias) {
+				return;
+			}
+
+			const resolvedAlias = await this.resolve(
+				unresolvedAlias,
+				import.meta.url
+			);
+			if (!resolvedAlias) {
+				return;
+			}
```

Comment:
Should we be asserting rather than returning `undefined` here? Otherwise, won't it resolve with the prefixed ID?

Reply:
I'm not sure the best practice here. If we return `undefined` then other plugins might have a crack at resolving it, right?
And then if nothing else can resolve this you'd get an error later on.

But if we are sure that failure to resolve here should be unrecoverable, we can throw instead.

Reply:
be3a0132eabda03a989eb68936ce748e6138f45f